x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUsers.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.user;
+
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class InternalUsers {
+
+    private record UserInstance(String name, User user, @Nullable RoleDescriptor role, Predicate<User> isUserPredicate) {
+        private UserInstance {
+            // Check that the parameters align
+            assert User.isInternal(user) : "User " + user + " is not internal";
+            // The user principal should match the official name
+            assert name.equals(user.principal())
+                : "User " + user + " has a principal [" + user.principal() + "] that dopes not match the provided name [" + name + "]";
```

Comment:
I was trying to match the existing code as much as possible to avoid subtle bugs and a hard-to-review diff.
   
 I'm happy to change it now if you'd like, but I'd prefer to just do a proper refactor on it all next week (🤞 )

Reply:
I don't expect there would be bugs by dropping the name here. But what you said is also true that we use both NAME and INSTANCE in places instead of only `INSTANCE` and `INSTANCE.principal()`. So I am OK with keeping it as is.